File: src/be/digitalia/fosdem/fragments/TrackScheduleListFragment.java
package be.digitalia.fosdem.fragments;

import java.text.DateFormat;

import android.app.Activity;
import android.content.Context;
import android.database.Cursor;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.v4.app.ListFragment;
import android.support.v4.app.LoaderManager.LoaderCallbacks;
import android.support.v4.content.Loader;
import android.support.v4.widget.CursorAdapter;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.AbsoluteSizeSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.TextView;
import be.digitalia.fosdem.R;
import be.digitalia.fosdem.db.DatabaseManager;
import be.digitalia.fosdem.loaders.TrackScheduleLoader;
import be.digitalia.fosdem.model.Day;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.Track;
import be.digitalia.fosdem.utils.DateUtils;

public class TrackScheduleListFragment extends ListFragment implements LoaderCallbacks<Cursor> {

	/**
	 * Interface implemented by container activities
	 */
	public interface Callbacks {
		void onEventSelected(int position, Event event);
	}

	private static final int EVENTS_LOADER_ID = 1;
	private static final String ARG_DAY = "day";
	private static final String ARG_TRACK = "track";

	private TrackScheduleAdapter adapter;
	private Callbacks listener;
	private boolean selectionEnabled = false;

	public static TrackScheduleListFragment newInstance(Day day, Track track) {
		TrackScheduleListFragment f = new TrackScheduleListFragment();
		Bundle args = new Bundle();
		args.putParcelable(ARG_DAY, day);
		args.putParcelable(ARG_TRACK, track);
		f.setArguments(args);
		return f;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		adapter = new TrackScheduleAdapter(getActivity());
		setListAdapter(adapter);
	}

	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		if (activity instanceof Callbacks) {
			listener = (Callbacks) activity;
		}
	}

	@Override
	public void onDetach() {
		super.onDetach();
		listener = null;
	}

	private void notifyEventSelected(int position) {
		if (listener != null) {
			listener.onEventSelected(position, (position == -1) ? null : adapter.getItem(position));
		}
	}

	public void setSelectionEnabled(boolean selectionEnabled) {
		this.selectionEnabled = selectionEnabled;
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		getListView().setChoiceMode(selectionEnabled ? ListView.CHOICE_MODE_SINGLE : ListView.CHOICE_MODE_NONE);
		setEmptyText(getString(R.string.no_data));
		setListShown(false);

		getLoaderManager().initLoader(EVENTS_LOADER_ID, null, this);
	}

	@Override
	public Loader<Cursor> onCreateLoader(int id, Bundle args) {
		Day day = getArguments().getParcelable(ARG_DAY);
		Track track = getArguments().getParcelable(ARG_TRACK);
		return new TrackScheduleLoader(getActivity(), day, track);
	}

	@Override
	public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
		if (data != null) {
			adapter.swapCursor(data);

			if (selectionEnabled) {
				final int count = adapter.getCount();
				int checkedPosition = getListView().getCheckedItemPosition();
				if ((checkedPosition == ListView.INVALID_POSITION) || (checkedPosition >= count)) {
					if (count > 0) {
						// Select the first item if any
						getListView().setItemChecked(0, true);
						checkedPosition = 0;
					} else {
						// No result, nothing selected
						checkedPosition = -1;
					}
				}

				// Ensure the current selection is visible
				if (checkedPosition != -1) {
					setSelection(checkedPosition);
				}
				// Notify the parent of the current selection to synchronize its state
				notifyEventSelected(checkedPosition);
			}
		}

		// The list should now be shown.
		if (isResumed()) {
			setListShown(true);
		} else {
			setListShownNoAnimation(true);
		}
	}

	@Override
	public void onLoaderReset(Loader<Cursor> loader) {
		adapter.swapCursor(null);
	}

	@Override
	public void onListItemClick(ListView l, View v, int position, long id) {
		notifyEventSelected(position);
	}

	private static class TrackScheduleAdapter extends CursorAdapter {

		private static final DateFormat TIME_DATE_FORMAT = DateUtils.getTimeDateFormat();

		private final LayoutInflater inflater;
		private final int titleTextSize;

		public TrackScheduleAdapter(Context context) {
			super(context, null, 0);
			inflater = LayoutInflater.from(context);
			titleTextSize = context.getResources().getDimensionPixelSize(R.dimen.list_item_title_text_size);
		}

		@Override
		public Event getItem(int position) {
			return DatabaseManager.toEvent((Cursor) super.getItem(position));
		}

		@Override
		public View newView(Context context, Cursor cursor, ViewGroup parent) {
			View view = inflater.inflate(R.layout.item_schedule_event, parent, false);

			ViewHolder holder = new ViewHolder();
			holder.time = (TextView) view.findViewById(R.id.time);
			holder.text = (TextView) view.findViewById(R.id.text);
			holder.titleSizeSpan = new AbsoluteSizeSpan(titleTextSize);
			holder.boldStyleSpan = new StyleSpan(Typeface.BOLD);
			view.setTag(holder);

			return view;
		}

		@Override
		public void bindView(View view, Context context, Cursor cursor) {
			ViewHolder holder = (ViewHolder) view.getTag();
			Event event = DatabaseManager.toEvent(cursor, holder.event);
			holder.event = event;
			holder.time.setText(TIME_DATE_FORMAT.format(event.getStartTime()));

			SpannableString spannableString;
			String eventTitle = event.getTitle();
			String personsSummary = event.getPersonsSummary();
			if (TextUtils.isEmpty(personsSummary)) {
				spannableString = new SpannableString(String.format("%1$s\n%2$s", eventTitle, event.getRoomName()));
			} else {
				spannableString = new SpannableString(String.format("%1$s\n%2$s\n%3$s", eventTitle, personsSummary, event.getRoomName()));
			}
			spannableString.setSpan(holder.titleSizeSpan, 0, eventTitle.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
			spannableString.setSpan(holder.boldStyleSpan, 0, eventTitle.length() + personsSummary.length() + 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

			holder.text.setText(spannableString);
			int bookmarkDrawable = DatabaseManager.toBookmarkStatus(cursor) ? R.drawable.ic_small_starred : 0;
			holder.text.setCompoundDrawablesWithIntrinsicBounds(0, 0, bookmarkDrawable, 0);
		}

		private static class ViewHolder {
			TextView time;
			TextView text;
			AbsoluteSizeSpan titleSizeSpan;
			StyleSpan boldStyleSpan;
			Event event;
		}
	}
}

File: app/src/main/java/be/digitalia/fosdem/fragments/TrackScheduleListFragment.kt
package be.digitalia.fosdem.fragments

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import be.digitalia.fosdem.R
import be.digitalia.fosdem.activities.TrackScheduleEventActivity
import be.digitalia.fosdem.adapters.TrackScheduleAdapter
import be.digitalia.fosdem.model.Day
import be.digitalia.fosdem.model.Track
import be.digitalia.fosdem.settings.UserSettingsProvider
import be.digitalia.fosdem.utils.assistedViewModels
import be.digitalia.fosdem.utils.getParcelableCompat
import be.digitalia.fosdem.utils.launchAndRepeatOnLifecycle
import be.digitalia.fosdem.viewmodels.TrackScheduleListViewModel
import be.digitalia.fosdem.viewmodels.TrackScheduleViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class TrackScheduleListFragment : Fragment(R.layout.recyclerview) {

    @Inject
    lateinit var userSettingsProvider: UserSettingsProvider
    @Inject
    lateinit var viewModelFactory: TrackScheduleListViewModel.Factory
    private val viewModel: TrackScheduleListViewModel by assistedViewModels {
        val args = requireArguments()
        val day: Day = args.getParcelableCompat(ARG_DAY)!!
        val track: Track = args.getParcelableCompat(ARG_TRACK)!!
        viewModelFactory.create(day, track)
    }
    private val activityViewModel: TrackScheduleViewModel by activityViewModels()
    private var isListAlreadyShown = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState != null) {
            isListAlreadyShown = savedInstanceState.getBoolean(STATE_IS_LIST_ALREADY_SHOWN)
        }
        selectedId = savedInstanceState?.getLong(STATE_SELECTED_ID)
            ?: requireArguments().getLong(ARG_FROM_EVENT_ID, RecyclerView.NO_ID)
    }

    private var selectedId: Long = RecyclerView.NO_ID

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(STATE_IS_LIST_ALREADY_SHOWN, isListAlreadyShown)
        outState.putLong(STATE_SELECTED_ID, selectedId)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val isSelectionEnabled = resources.getBoolean(R.bool.tablet_landscape)

        val adapter = TrackScheduleAdapter(view.context) { event ->
            selectedId = event.id
            activityViewModel.selectedEvent = event

            if (!isSelectionEnabled) {
                // Classic mode: Show event details in a new activity
                val intent = Intent(requireContext(), TrackScheduleEventActivity::class.java)
                    .putExtra(TrackScheduleEventActivity.EXTRA_DAY, event.day)
                    .putExtra(TrackScheduleEventActivity.EXTRA_TRACK, event.track)
                    .putExtra(TrackScheduleEventActivity.EXTRA_EVENT_ID, event.id)
                startActivity(intent)
            }
        }
        val holder = RecyclerViewViewHolder(view).apply {
            recyclerView.apply {
                layoutManager = LinearLayoutManager(context)
                addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
            }
            setAdapter(adapter)
            emptyText = getString(R.string.no_data)
            isProgressBarVisible = true
        }

        viewLifecycleOwner.launchAndRepeatOnLifecycle {
            launch {
                userSettingsProvider.zoneId.collect { zoneId ->
                    adapter.zoneId = zoneId
                }
            }

            launch {
                viewModel.schedule.collect { schedule ->
                    adapter.submitList(schedule)

                    var selectedPosition = if (selectedId == -1L) -1 else schedule.indexOfFirst { it.event.id == selectedId }
                    if (selectedPosition == -1) {
                        // There is no current valid selection, reset to use the first item (if any)
                        if (schedule.isNotEmpty()) {
                            selectedPosition = 0
                            selectedId = schedule[0].event.id
                        } else {
                            selectedId = -1L
                        }
                    }

                    activityViewModel.selectedEvent =
                        if (selectedPosition == -1) null else schedule[selectedPosition].event

                    // Ensure the selection is visible
                    if ((isSelectionEnabled || !isListAlreadyShown) && selectedPosition != -1) {
                        holder.recyclerView.scrollToPosition(selectedPosition)
                    }
                    isListAlreadyShown = true

                    holder.isProgressBarVisible = false
                }
            }

            launch {
                viewModel.currentTime.collect { now ->
                    adapter.currentTime = now
                }
            }

            if (isSelectionEnabled) {
                launch {
                    activityViewModel.selectedEventFlow.collect { event ->
                        adapter.selectedId = event?.id ?: RecyclerView.NO_ID
                    }
                }
            }
        }
    }

    companion object {
        private const val ARG_DAY = "day"
        private const val ARG_TRACK = "track"
        private const val ARG_FROM_EVENT_ID = "from_event_id"

        private const val STATE_IS_LIST_ALREADY_SHOWN = "isListAlreadyShown"
        private const val STATE_SELECTED_ID = "selectedId"

        fun createArguments(day: Day, track: Track, fromEventId: Long = RecyclerView.NO_ID) = Bundle(3).apply {
            putParcelable(ARG_DAY, day)
            putParcelable(ARG_TRACK, track)
            putLong(ARG_FROM_EVENT_ID, fromEventId)
        }
    }
}